cdk/PgBouncer.ts

```diff
@@ -0,0 +1,179 @@
+import {
+  aws_ec2 as ec2,
+  aws_iam as iam,
+  aws_secretsmanager as secretsmanager,
+  Stack,
+} from "aws-cdk-lib";
+import { Construct } from "constructs";
+
+import * as fs from "fs";
+import * as path from "path";
+
+export interface PgBouncerProps {
+  /**
+   * Name for the pgbouncer instance
+   */
+  instanceName: string;
+
+  /**
+   * VPC to deploy PgBouncer into
+   */
+  vpc: ec2.IVpc;
+
+  /**
+   * The RDS instance to connect to
+   */
+  database: {
+    connections: ec2.Connections;
+    secret: secretsmanager.ISecret;
+  };
+
+  /**
+   * Whether to deploy in public subnet
+   * @default false
+   */
+  usePublicSubnet?: boolean;
+
+  /**
+   * Instance type for PgBouncer
+   * @default t3.small
+   */
+  instanceType?: ec2.InstanceType;
+
+  /**
+   * PgBouncer configuration options
+   */
+  pgBouncerConfig: {
+    poolMode: "transaction" | "session" | "statement";
+    maxClientConn: number;
+    defaultPoolSize: number;
+    minPoolSize: number;
+    reservePoolSize: number;
+    reservePoolTimeout: number;
+    maxDbConnections: number;
+    maxUserConnections: number;
+  };
+}
+
+export class PgBouncer extends Construct {
+  public readonly instance: ec2.Instance;
+  public readonly endpoint: string;
+
+  constructor(scope: Construct, id: string, props: PgBouncerProps) {
+    super(scope, id);
+
+    const {
+      vpc,
+      database,
+      usePublicSubnet = false,
+      instanceType = ec2.InstanceType.of(
+        ec2.InstanceClass.T3,
+        ec2.InstanceSize.MICRO,
+      ),
+      pgBouncerConfig = {
+        poolMode: "transaction",
+        maxClientConn: 200,
+        defaultPoolSize: 5,
+        minPoolSize: 0,
+        reservePoolSize: 5,
+        reservePoolTimeout: 5,
+        maxDbConnections: 40,
```

Comment:
Do these connection default configs also follow the same "pattern" as `const maxPgBouncerDbConnections`?

Reply:
Yes it could go into the construct! I think the only parameter that we need to adjust based on instance type is `maxDbConnections`. The rest of these parameters influence how the application handles connections on the pgbouncer server but are not relevant to the size of the db instance. I'll add some notes here to highlight the important details!

Reply:
I remember why I left it out of the construct originally. Maybe we could fall back to this logic as a default in the `Pgbouncer` construct, but I want to export the `max_connections` setting in the `PgstacDatabase` construct in eoapi-cdk. That way we can stay consistent with the database settings in case they are not following this rough calculation.